### ISSUE TITLE
STI: class names as parameters for where

### ISSUE THREAD

**Issue by skyfmmf** [rails/rails#31100]
### Steps to reproduce
While upgrading Rails to 5.1.4 I noticed the following behaviour: single class names are no longer accepted as parameters for where, but they are accepted when enclosed in an array. This makes it more difficult to use the inheritance column in queries. 

I hope the following examples show what I mean: 

```ruby
Animal.where(:type => Cat) # broken in rails 5.1.4
Animal.where(:type => "Cat") # works, but misses the check for existence of the class on autoload
Animal.where(:type => [Cat]) # works, but constructs a LIKE query
```

### Expected behavior
I expected the where to work with class names as parameters for the inheritance column like before. 

### Actual behavior
I get a `TypeError: can't cast Class` when executing the query.

### System configuration
**Rails version**: 5.1.4

**Ruby version**: 2.3.3

**Comment by pixeltrix** [comment]
@skyfmmf is this a regression or has it always been this way?

**Comment by pixeltrix** [comment]
@skyfmmf I can't reproduce this in Rails 5.1.4 - looking at the error message it doesn't appear to recognise the column but even that shouldn't cause a problem. If you look at the code for quoting values it has `Class` in there:

``` ruby
def _quote(value)
  case value
  when String, ActiveSupport::Multibyte::Chars
    "'#{quote_string(value.to_s)}'"
  when true       then quoted_true
  when false      then quoted_false
  when nil        then "NULL"
  # BigDecimals need to be put in a non-normalized form and quoted.
  when BigDecimal then value.to_s("F")
  when Numeric, ActiveSupport::Duration then value.to_s
  when Type::Binary::Data then quoted_binary(value)
  when Type::Time::Value then "'#{quoted_time(value)}'"
  when Date, Time then "'#{quoted_date(value)}'"
  when Symbol     then "'#{quote_string(value.to_s)}'"
  when Class      then "'#{value}'"
  else raise TypeError, "can't quote #{value.class.name}"
  end
end
```

Looking at the error you're getting it seems like it's happening after the query has been executed and it's trying to initialize the model instances - perhaps there's something that's causing the error there? Please try to create a reproducible test case [using the template][1], thanks.

[1]: https://raw.githubusercontent.com/rails/rails/master/guides/bug_report_templates/active_record_gem.rb

**Comment by skyfmmf** [comment]
Thanks for your fast response. 

I am doing an upgrade of rails so I think it is an regression. I could reproduce the bug using the template, I could also verify it worked in version 4.2.7 and does not work with 5.1.4. You can find my test here: https://gist.github.com/skyfmmf/108bcf5ac4ea0477fd77a4a3d17c400c

**Comment by pixeltrix** [comment]
@skyfmmf thanks for the test - I can see where the problem is. We're typecasting the binds to database values for prepared statements and instrumentation but it doesn't handle objects of type `Class`. This is happening even though the SQL is built correctly - you can see the first `SELECT` query in the log output. I suspect the issue started with Rails 5.0.

**Comment by matthewd** [comment]
Wasn't this deprecated and removed?

**Comment by rafaelfranca** [comment]
Yes

**Comment by rafaelfranca** [comment]
https://github.com/rails/rails/commit/b4664864c972463c7437ad983832d2582186e886

**Comment by skyfmmf** [comment]
Oh, I did not see this. 

Just curious: why was this removed? It seems like the best way to make queries with the type in STI. The most important part of this for me was the check for existence of the class due to autoloading. It would fail early with a useful error message. To retain this feature I had to convert every `.where(type: TextPost)` to `.where(type: TextPost.to_s)` what seems not that nice to me. 

Also the removal was not in all places. Queries like `.where(type: [TextPost, ImagePost]` do still work without converting the class to a string. Will this also be removed in future releases? Or will this inconsistency be kept?

**Comment by pixeltrix** [comment]
@rafaelfranca @matthewd should we be raising in `_quote` then rather than building the SQL and then raising after it's been executed?

**Comment by rafaelfranca** [comment]
I'm not sure. I feel checking for types is too defensive. We do in some places so it is fine to me

**Comment by rafaelfranca** [comment]
https://github.com/rails/rails/pull/17916

**Comment by rafaelfranca** [comment]
@skyfmmf https://github.com/rails/rails/pull/17916 is the PR that deprecated it

**Comment by matthewd** [comment]
Yeah, I'm not sure either -- particularly whether `_quote` should still have special handling for Class.

To me, the compelling argument for deprecating `where(x: Foo)` in particular (that is, its use in a where condition) is that it would be natural to read that as including Foo's descendants. I do think the where+array form should therefore go away (if it hasn't already due to refactorings in master); the fact that currently works seems to be simple oversight.

That argument is less strong in other places, where it's much less ambiguous that we're converting the supplied value to a single SQL value... but at the same time, allowing the value there is inconsistent, and seems to offer little benefit (given you can't use it in the more ergonomic where-hash form).

cc @sgrif 

----

> It seems like the best way to make queries with the type in STI. The most important part of this for me was the check for existence of the class due to autoloading. It would fail early with a useful error message.

This part I don't follow... autoloading should.. autoload the class. So I'm unclear what you're trying to "check for existence".

The best way to make queries with the type in STI should be to query with the type: `TextPost.first` etc. Ideally, you should need to directly touch the `type` column about as often as you do `id` -- i.e., not very much at all.

**Comment by sgrif** [comment]
Yeah I think we can probably get rid of the handling in `_quote`, that seems like it was just an oversight and the fact that the array form continued to work was an unintentional oversight. If we were to do anything like Matthew proposed, that'd get handled much earlier in the predicate builder.

**Comment by skyfmmf** [comment]
@matthewd 

> This part I don't follow... autoloading should.. autoload the class. So I'm unclear what you're trying to "check for existence".

Yes autoloading should autoload the class, but it fails when class can not be found (i.e. because it has been renamed or there is a typo in the name). When the class name given as a string it will fail when reaching the code. Putting the class name in there as a constant triggers the autoloader, which fails loading the class before executing the code.

**Comment by matthewd** [comment]
I still don't understand.. they both happen when that line of code is executed. (One is before the query and one is after, but I'm unsure how that distinction would matter.)

**Comment by skyfmmf** [comment]
When working with an huge codebase and eager loading activated even in development for various reasons, the one with the class name fails at the startup of the rails server

**Comment by msxavi** [comment]
Hey guys, how is this issue going? You're not going to revert this change, are you? Just to make sure.

**Comment by dmitry** [comment]
I believe there should be something like:

```
when ActiveRecord::Base      then "'#{value.to_s}'"
```

Came here after the MySQL -> PostgreSQL migration. MySQL adapter works fine with such syntax.

**Comment by marcelolx** [comment]
Based on the conversation I don't think that this was considered a bug since the behavior was removed intentionally.

**Comment by gregorw** [comment]
I’d opt for
```ruby
when BasicObject then "'#{value}'"
```

because 

```ruby
irb(main):001> class Test; end
=> nil
irb(main):002> Class === Test.new
=> false
irb(main):003> BasicObject === Test.new
=> true
```

**Comment by matthewd** [comment]
@gregorw you're talking about making _every_ object silently work in any query. That's quite unlike the focus of discussion in this (ancient) issue, and also definitely not something we're going to do.

You can make your example work by defining a Type for the `color` attribute in the Product model.

**Comment by janko** [comment]
> Came here after the MySQL -> PostgreSQL migration. mysql2 adapter works fine with such syntax and for the postgresql I have to add .to_s everywhere.

Yeah, I was surprised to have this working on MySQL adapter, but when I tried to make our tests pass on Postgres, I got the cryptic `TypeError: can't cast Class` that took me a while to figure out.

If this shouldn't work, then I would expect it not to work on all adapters.
